hazelcast/src/main/java/com/hazelcast/jet/core/Processor.java

```diff
@@ -184,32 +184,126 @@ default void process(int ordinal, @Nonnull Inbox inbox) {
+     * the watermark in any way: drop it, delay it or move it ahead, change the
+     * key, or even emit a completely different watermark, as long as the output
+     * watermarks are monotonic. Any processing method can emit watermarks. Sink
+     * processors in general should ignore the watermark and simply return
+     * {@code true}.
+     *
+     * <h3>Difference between the overloaded {@code tryProcessWatermarks()}
+     * variants</h3>
+     *
+     * The method is available in two overloaded variants: with and without an
+     * edge ordinal:<ul>
+     *     <li>The variant <b>with ordinal</b> is called after the watermark was
+     *     received from all upstream processors contributing to that input
+     *     ordinal.
+     *     <li>The variant <b>without an ordinal</b> is called after the
+     *     watermark was received from all input ordinals.
+     * </ul>
+     *
+     * Which method to override depends on the purpose of the processor. For
+     * example, a join processor can receive different watermark from each
+     * input, so it needs to override the variant with the ordinal. A merging
+     * processor, on the other hand, expects the same watermarks from all
+     * inputs, so it overrides the variant without an ordinal. Each watermark is
+     * passed to both methods, so in most cases you need to override at most one
+     * method. However, if a watermark with some key is received only from one
+     * input edge, the variant without the ordinal is never called for that
+     * watermark.
+     * <p>
+     * Also, please, pay attention to the default implementations in this class,
+     * and in {@link AbstractProcessor}, which handle the case of merging
+     * streams.
      *
      * <h3>Caution for Jobs With the At-Least-Once Guarantee</h3>
+     *
      * Jet propagates the value of the watermark by sending <em>watermark
-     * items</em> interleaved with the regular stream items. If a job
-     * configured with the <i>at-least-once</i> processing guarantee gets
-     * restarted, the same watermark, like any other stream item, can be
-     * delivered again. Therefore the processor may be asked to process
-     * a watermark older than the one it had already processed before the
-     * restart.
+     * items</em> interleaved with the regular stream items. If a job configured
+     * with the <i>at-least-once</i> processing guarantee gets restarted, the
+     * same watermark, like any other stream item, can be delivered again.
+     * Therefore, the processor may be asked to process a watermark older than
+     * the one it had already processed before the restart.
      *
      * @param watermark watermark to be processed
      * @return {@code true} if this watermark has now been processed,
      *         {@code false} to call this method again with the same watermark
      */
     boolean tryProcessWatermark(@Nonnull Watermark watermark);
```

Comment:
We should update the processors that need to take the key into account. If we don't do it in this PR, we must add a check to them to fail early if `key != 0` (not an assert).